### PR TITLE
*: support generates metrics profile by different type

### PR DESCRIPTION
**Pull request by crazycs520**
### What problem does this PR solve?

PR https://github.com/pingcap/tidb/pull/18272 supports to generated a metric profile, but there are several defects, this PR try to make it better to use.

1. support generates metrics profile by different type, such as generate the profile by avg time and total count too.

Add a `type` parameter in HTTP API  to specified the profile type, expect value is one of `sum` of  `avg` or `count`, default type is `sum`. Usage example:

```sql
curl "127.0.0.1:10080/metrics/profile?type=avg&start=2020-07-23T19:44:00%2B08:00&end=2020-07-23T19:49:00%2B08:00" > profile.dot
```

image

2. add a comment for each node. Display more value in tips. Such as:  
* `total_count`
* `avg_time`
* `total_time`
* `P99`
* `P90`
* `P80`

image

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Manual test

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

- support generates metrics profile by different type

### CODE REVIEW

File: executor/inspection_profile.go [hunk 1]
```diff
@@ -34,22 +39,96 @@ type profileBuilder struct {
...
+	count  int
+	avgP99 float64
+	avgP90 float64
+	avgP80 float64
+}
+
+type metricValueType int
+
+const (
+	metricValueSum metricValueType = 1
```
Comment (wjhuang2016) [inline executor/inspection_profile.go:70]: use itoa ?

File: executor/inspection_profile.go [hunk 2]
```diff
@@ -61,71 +140,162 @@ func (n *metricNode) getName(label string) string {
...
+func (n *metricNode) initializeMetricValue(pb *profileBuilder) error {
+	n.labelValue = make(map[string]*metricValue)
+	n.value = &metricValue{}
+	queryCondition := fmt.Sprintf("where time >= '%v' and time <= '%v' and value is not null and value>0",
+		pb.start.Format(dateTimeFormat), pb.end.Format(dateTimeFormat))
 	if n.condition != "" {
 		queryCondition += (" and " + n.condition)
 	}
+
+	queryRows := func(query string, fn func(label string, v float64)) error {
```
Comment (wjhuang2016) [inline executor/inspection_profile.go:172]: The code here is really hard to read, please use the normal function.

File: executor/inspection_profile.go (hunk 1 above)
Comment (crazycs520) [reply executor/inspection_profile.go]: done.

File: executor/inspection_profile.go (hunk 2 above)
Comment (crazycs520) [reply executor/inspection_profile.go]: done.

File: executor/inspection_profile.go [hunk 3]
```diff
@@ -118,14 +184,118 @@ func (pb *profileBuilder) getMetricValue(n *metricNode) (float64, error) {
...
+		valueTp = metricValueSum
+	case metricValueAvg.String():
+		valueTp = metricValueAvg
+	case metricValueCnt.String():
+		valueTp = metricValueCnt
+	case "":
+		// Use type sum when doesn't specified the type, this is used to compatible with old behaviour.
+		valueTp = metricValueSum
+	default:
+		return nil, fmt.Errorf("unknown metric profile type: %v, expect value should be one of sum or avg or count", tp)
```
Comment (tangenta) [inline executor/inspection_profile.go:297]:
```suggestion
		return nil, fmt.Errorf("unknown metric profile type: %v, expect value should be one of 'sum', 'avg' or 'count'", tp)
```

**Pull request merged**